### ISSUE TITLE
Access to underlying filename

### ISSUE THREAD

**Issue by kowitz** [superwolff/metalsmith-layouts#58]
I'm trying to render markdown files using a handlebars layout. I need access to the original markdown filename. But unfortunately, the `filename` template variable is showing the name of the layout, instead of the underlying markdown file. 

Here's my config from the CLI:

```
          'metalsmith-layouts': {
            engine: 'handlebars',
            partials: 'partials',
            directory: 'layouts',
            pattern: 'team/*',
            default: 'layout-bio.html'
          },
```

And in this case, when I use `{{filename}}`, is equals `layouts/layout-bio.html` instead of something like `team/braden-kowitz.md`

I'm a bit stuck, so any suggestions would be greatly appreciated!
Thanks.

**Comment by ismay** [comment]
Hmm, I see. Just so I have some context, what's the problem you're trying to solve? There might be an easier way.

**Comment by kowitz** [comment]
Sure!

I'm working on a team page for a website. I'm trying to render a folder of personal bios written in markdown format into html pages on the final site. So, for example, there's a file called `team/braden-kowitz.md` that needs to be rendered by the layout `layouts/layout-bio.html`.

In that layout file, I use frontmatter from the markdown file to pull in info about the person, for example `{{name}}` and `{{title}}`. I also want to include a photo of the person, which shares the same slug from the filename. So if the file is `braden-kowitz.md`, I'd like to point to an image file like `/img/team/braden-kowitz.jpg`.

So I need access to the filename in order to link to the profile image correctly. Does that makes sense? 

Thanks for the help!
- Braden

**Comment by ismay** [comment]
> So I need access to the filename in order to link to the profile image correctly. Does that makes sense? 

Ok I see, that makes sense. I think the easiest thing to do is use [metalsmith-collections](https://github.com/segmentio/metalsmith-collections). I use it myself [here](https://github.com/ismay/ismaywolff.nl/blob/master/build.js#L132) and then loop through the generated collection [here](https://github.com/ismay/ismaywolff.nl/blob/master/src/index.swig#L12). I think that should work similar to what you're trying to do, if I'm not mistaken.

It'll probably be easier to do it that way. I am using [metalsmith-in-place](https://github.com/superwolff/metalsmith-in-place) though (which is pretty similar), so you'll might have to adapt some things if you want to keep using layouts.

**Comment by kowitz** [comment]
I'm not sure that metalsmith-collections will help actually. I don't need to render an index page listing the team members. I just need each markdown filename to be accessible while the layout is rendering.

For example, I have a `jane-smith.md` file that looks something like this:
```
--- 
name: Jane Smith
title: President
---
Jane Smith started her career as a javascript developer.
Now she's president of ACME Corp.
```

When this markdown file is run through the layout, I'd like to have access to the name `jane-smith.md`, but it seems to be overwritten by the name of the layout file. I'm not sure what the proper behavior is here, but it seems like a bug. Shouldn't `{{filename}}` match to the file being rendered, not the layout?

Thanks for the help here! The rest of my experience with Metalsmith has been amazing. This is just the one little hiccup that I'm trying to figure out still.

Cheers,
— Braden

**Comment by ismay** [comment]
> I'm not sure that metalsmith-collections will help actually. I don't need to render an index page listing the team members. I just need each markdown filename to be accessible while the layout is rendering.

Yeah ok, so you just want the filename of the markdown file that's currently being rendered into a template. And I assume that you're using `metalsmith-filenames`?

Metalsmith filenames makes the filename of each file available as a variable on the object associated with that file. What you're running into is a limitation of `metalsmith-layouts`, which only processes variables in the layout, and not from the contents being passed to the layout (the markdown file). Hence, the filename isn't available.

You could use `metalsmith-in-place`, and apply templates by extending them, but I don't think you'd be able to use markdown. If I were you I'd switch to `metalsmith-in-place`, and use it without the markdown plugin. That way you'll be able to do what you're trying to do here.

There are probably other ways as well, but they'll be a lot more cumbersome and difficult to implement. Hope that helps. If not, an example repo might help illustrate what you're trying to do. Let me know if you have any questions!

**Comment by kowitz** [comment]
Yeah, I was using `metalsmith-filenames`, but it seems that the layout module overwrites the filename attribute. I didn't want to use `metalsmith-in-place` because we have dozens of markdown bio files and I want to keep application logic out of those files. 

I did figure it out though! My solution was to create my own plugin that takes the file's name (minus the extension) and writes it to an attribute called `slug`. In case anyone is in this situation, here's what I did:

First, I created a plugin at `/metalsmith/slugs.js`
```javascript
var path = require('path');
module.exports = function() {
  return function(files, metalsmith, callback) {
    Object.keys(files).forEach(function(filename){
      files[filename].slug = path.parse(filename).name;
    });
    callback();
  };
};
```

Then, I included it in the CLI object. I'm using `metalsmith-grunt`, so this is part of my `Gruntfile.js`:
```javascript
plugins: {
    '../../../metalsmith/slugs.js': {},
    // ... other plugins
}
```

@ismay, Thanks for all the help with this. I appreciate the encouragement.
Cheers,
— Braden
